crates/factor-outbound-networking/src/lib.rs

```diff
@@ -0,0 +1,116 @@
+use std::{collections::HashMap, sync::Arc};
+
+use anyhow::Context;
+use futures_util::{
+    future::{BoxFuture, Shared},
+    FutureExt,
+};
+use spin_factor_variables::VariablesFactor;
+use spin_factor_wasi::WasiFactor;
+use spin_factors::{
+    Factor, FactorInstancePreparer, InstancePreparers, PrepareContext, Result, SpinFactors,
+};
+use spin_outbound_networking::{AllowedHostsConfig, ALLOWED_HOSTS_KEY};
+
+pub struct OutboundNetworkingFactor;
+
+impl Factor for OutboundNetworkingFactor {
+    type AppConfig = AppConfig;
+    type InstancePreparer = InstancePreparer;
+    type InstanceState = ();
+
+    fn configure_app<Factors: SpinFactors>(
+        &self,
+        app: &spin_factors::App,
+        _ctx: spin_factors::ConfigureAppContext<Factors>,
+    ) -> Result<Self::AppConfig> {
+        // Extract allowed_outbound_hosts for all components
+        let component_allowed_hosts = app
+            .components()
+            .map(|component| {
+                Ok((
+                    component.id().to_string(),
+                    component
+                        .get_metadata(ALLOWED_HOSTS_KEY)?
+                        .unwrap_or_default()
+                        .into_boxed_slice()
+                        .into(),
+                ))
+            })
+            .collect::<Result<_>>()?;
+        Ok(AppConfig {
+            component_allowed_hosts,
+        })
+    }
+}
+
+#[derive(Default)]
+pub struct AppConfig {
+    component_allowed_hosts: HashMap<String, Arc<[String]>>,
+}
+
+type AllowedHostsFuture = Shared<BoxFuture<'static, Arc<anyhow::Result<AllowedHostsConfig>>>>;
+
+pub struct InstancePreparer {
+    allowed_hosts_future: AllowedHostsFuture,
+}
+
+impl FactorInstancePreparer<OutboundNetworkingFactor> for InstancePreparer {
+    fn new<Factors: SpinFactors>(
+        ctx: PrepareContext<OutboundNetworkingFactor>,
+        mut preparers: InstancePreparers<Factors>,
+    ) -> Result<Self> {
+        let hosts = ctx
+            .app_config()
+            .component_allowed_hosts
+            .get(ctx.app_component().id())
+            .cloned()
+            .context("missing component allowed hosts")?;
+        let resolver = preparers.get_mut::<VariablesFactor>()?.resolver().clone();
```

Comment:
I think this would make the potential error message much clearer, but it would also make the code easier to understand:

```suggestion
        let resolver = preparers.get_mut::<VariablesFactor>().context("ADD HELPFUL MESSAGE HERE")?.resolver().clone();
```

Reply:
`get_mut` already has `.context("preparer not initialized")`, but also there are a few `anyhow::Result`s that I intend to replace with `Result<_, spin_factors::Error>` once the overall structure of things has settled down a bit.